packages/checkout/widgets-lib/src/lib/hooks/useWalletConnect.ts

```diff
@@ -0,0 +1,97 @@
+import { useCallback, useEffect, useState } from 'react';
+import EthereumProvider from '@walletconnect/ethereum-provider';
+import { WalletConnectModal } from '@walletconnect/modal';
+import { WalletConnectManager } from '../walletConnect';
+
+export interface UseWalletConnectParams {
+  connectCallback: (ethereumProvider: EthereumProvider) => void,
+}
+
+export const useWalletConnect = ({ connectCallback }: UseWalletConnectParams) => {
+  const [walletConnectBusy, setWalletConnectBusy] = useState<boolean>(false);
+  const [ethereumProvider, setEthereumProvider] = useState<EthereumProvider | null>(null);
+  const [walletConnectModal, setWalletConnectModal] = useState<WalletConnectModal | null>(null);
+  // const [displayUri, setDisplayUri] = useState<string | null>(null);
+
+  useEffect(() => {
+    (async () => setEthereumProvider(await WalletConnectManager.getInstance().getProvider()))();
+    setWalletConnectModal(WalletConnectManager.getInstance().getModal());
+  }, []);
+
+  const openWalletConnectModal = useCallback(async (restoreSession: boolean = true) => (
+    new Promise((resolve, reject) => {
+      if (!ethereumProvider || !walletConnectModal) reject('WalletConnect not initialized');
+      setWalletConnectBusy(true);
+
+      if (restoreSession) {
+        // restore session
+        const existingPairings = ethereumProvider?.signer.client.core.pairing.getPairings();
+        console.log('existingPairings', existingPairings);
+        if (existingPairings && existingPairings.length > 0 && existingPairings[0].topic !== '') {
+          ethereumProvider?.signer.client.core.pairing.activate({ topic: existingPairings[0].topic })
+            .then(() => {
+              console.log('restored session');
+              if (connectCallback) {
+                connectCallback(ethereumProvider);
+              }
+            });
+          return;
+        }
+      }
+
+      // Hook into next available display_uri
+      ethereumProvider?.once('display_uri', (data) => {
+        // eslint-disable-next-line no-console
+        console.log('** hook wc display_uri', data);
+        // const pairingTopic = data.split('@')[0].replace('wc:', '');
+        // // eslint-disable-next-line no-console
+        // console.log('** hook pairingTopic', pairingTopic);
+
+        // setDisplayUri(data);
+        walletConnectModal?.openModal({
+          uri: data,
+        })
+          .then((result) => {
+            setWalletConnectBusy(false);
+            resolve(result);
+          })
+          .catch((error) => {
+            // Error opening WalletConnect Modal
+            setWalletConnectBusy(true);
+            reject(error);
+          });
+      });
+
+      ethereumProvider?.once('connect', (data) => {
+        console.log(data);
+        walletConnectModal?.closeModal();
+
+        if (connectCallback) {
+          connectCallback(ethereumProvider);
+        }
+      });
```

Comment:
Can we test if this event is fired when connecting MetaMask mobile twice (but not through the restore session method, just two separate connections)?